### ISSUE TITLE
dns failures on 1.9.1 inside existing vpc

### ISSUE THREAD

**Issue by inodb** [kubernetes/kops#5236]
1. What `kops` version are you running? The command `kops version`, will display
 this information.
```
$ kops version
Version 1.9.1
```

2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.2", GitCommit:"81753b10df112992bf51bbc2c2f85208aad78335", GitTreeState:"clean", BuildDate:"2018-05-12T04:12:47Z", GoVersion:"go1.9.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.6", GitCommit:"9f8ebd171479bec0ada837d7ee641dec2f8c6dd1", GitTreeState:"clean", BuildDate:"2018-03-21T15:13:31Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

3. What cloud provider are you using?
AWS

4. What commands did you run?  What is the simplest way to reproduce this issue?
Created a cluster within existing empty aws default vpc
```
export VPC_ID=xxx
export SUBNET_IDS=subnet-xxx
export KOPS_STATE_STORE=s3://xxxx.k8s.local
export NAME=xxxx.k8s.local

kops create cluster --state=${KOPS_STATE_STORE} --cloud=aws  --zones=us-east-1a --node-count=2  --node-size=t2.medium --master-size=t2.small ${NAME} --subnets=${SUBNET_IDS}
```
After waiting for the cluster, got two nodes as expected:
```
$ kubectl get nodes --show-labels
NAME                            STATUS    ROLES     AGE       VERSION   LABELS
ip-172-31-82-164.ec2.internal   Ready     master    22m       v1.9.6    beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=t2.small,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1a,kops.k8s.io/instancegroup=master-us-east-1a,kubernetes.io/hostname=ip-172-31-82-164.ec2.internal,kubernetes.io/role=master,node-role.kubernetes.io/master=
ip-172-31-91-241.ec2.internal   Ready     node      21m       v1.9.6    beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=t2.medium,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1a,kops.k8s.io/instancegroup=nodes,kubernetes.io/hostname=ip-172-31-91-241.ec2.internal,kubernetes.io/role=node,node-role.kubernetes.io/node=
ip-172-31-94-234.ec2.internal   Ready     node      22m       v1.9.6    beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=t2.medium,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-east-1,failure-domain.beta.kubernetes.io/zone=us-east-1a,kops.k8s.io/instancegroup=nodes,kubernetes.io/hostname=ip-172-31-94-234.ec2.internal,kubernetes.io/role=node,node-role.kubernetes.io/node=
```
However it seems like DNS isn't working. nslookup in busybox fails for both nodes:
```
$ kubectl delete po busybox; kubectl run -i --tty busybox --image=busybox --restart=Never --overrides='{ "apiVersion": "v1", "spec": { "nodeSelector": {"kubernetes.io/hostname":"ip-172-31-91-241.ec2.internal"} } }' -- nslookup kubernetes.default
pod "busybox" deleted
If you don't see a command prompt, try pressing enter.
Address 1: 100.64.0.10 kube-dns.kube-system.svc.cluster.local
nslookup: can't resolve 'kubernetes.default'
pod default/busybox terminated (Error)
```

```
$ kubectl delete po busybox; kubectl run -i --tty busybox --image=busybox --restart=Never --overrides='{ "apiVersion": "v1", "spec": { "nodeSelector": {"kubernetes.io/hostname":"ip-172-31-94-234.ec2.internal"} } }' -- nslookup kubernetes.default
pod "busybox" deleted
If you don't see a command prompt, try pressing enter.
Address 1: 100.64.0.10

nslookup: can't resolve 'kubernetes.default'
pod default/busybox terminated (Error)
```

Sporadically the command succeeds:
```
Name:      kubernetes.default
Address 1: 100.64.0.1 kubernetes.default.svc.cluster.local
```

5. What happened after the commands executed?
See above, dns does not seem to be working. Sporadically a succesful

6. What did you expect to happen?
nslookup should work

7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -oyaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.
```
apiVersion: kops/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: 2018-05-31T20:53:25Z
  name: xxx.k8s.local
spec:
  api:
    loadBalancer:
      type: Public
  authorization:
    rbac: {}
  channel: stable
  cloudProvider: aws
  configBase: s3://xxx/xxx.k8s.local
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master-us-east-1a
      name: a
    name: main
  - etcdMembers:
    - instanceGroup: master-us-east-1a
      name: a
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.9.6
  masterPublicName: xxx.k8s.local
  networkCIDR: 172.31.0.0/16
  networkID: xxx
  networking:
    kubenet: {}
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 0.0.0.0/0
  subnets:
  - cidr: 172.31.80.0/20
    id: xxx
    name: us-east-1a
    type: Public
    zone: us-east-1a
  topology:
    dns:
      type: Public
    masters: public
    nodes: public

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-05-31T20:53:25Z
  labels:
    kops.k8s.io/cluster: xxx.local
  name: master-us-east-1a
spec:
  image: kope.io/k8s-1.9-debian-jessie-amd64-hvm-ebs-2018-03-11
  machineType: t2.small
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-us-east-1a
  role: Master
  subnets:
  - us-east-1a

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-05-31T20:53:25Z
  labels:
    kops.k8s.io/cluster: xxx.k8s.local
  name: nodes
spec:
  image: kope.io/k8s-1.9-debian-jessie-amd64-hvm-ebs-2018-03-11
  machineType: t2.medium
  maxSize: 2
  minSize: 2
  nodeLabels:
    kops.k8s.io/instancegroup: nodes
  role: Node
  subnets:
  - us-east-1a
```

**Comment by inodb** [comment]
☝️ as a side note, which should prolly be a separate issue. I think `kops validate` should catch these dns failures

**Comment by jueast08** [comment]
I'm running into a similar problem. Have you been able to figure this one out ?

**Comment by inodb** [comment]
@jueast08 unf not - it was fine in my case to create a new vpc instead so settled for that solution

**Comment by lindenle** [comment]
I am having the same issue.

**Comment by vsayer** [comment]
Also having the same issue

```
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.0", GitCommit:"0ed33881dc4355495f623c6f22e7dd0b7632b7c0", GitTreeState:"clean", BuildDate:"2018-09-28T15:20:58Z", GoVersion:"go1.11", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.6", GitCommit:"a21fdbd78dde8f5447f5f6c331f7eb6f80bd684e", GitTreeState:"clean", BuildDate:"2018-07-26T10:04:08Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by danielgrant** [comment]
I was struggling with this exact same issue and finally cracked it. In my case, Kubernetes was unable to determine which AWS route table to interact with to create routes on-the-fly.

Check the output of the `kub-controller-manager-ip` pod in the cluster:

```
kubectl get pods --namespace=kube-system
```

Locate the correct pod and check the logs:

```
kubectl logs kube-controller-manager-ip-... --namespace=kube-system
```

I had this error being generated over and over:

```
E0111 17:35:15.121422       1 route_controller.go:117] Couldn't reconcile node routes: error listing routes: found multiple matching AWS route tables for AWS cluster: kubernetes.example.com
```

For me, it was because I had multiple route tables defined in AWS with the tag `KubernetesCluster` and value `kubernetes.example.com` (one route table for the public subnets and one for private subnets). As soon as I ensured that my master and node instances were in the public subnets only and removed the `KubernetesCluster` tag from the private subnet, the problem immediately went away.

Hope that helps.

**Comment by vsayer** [comment]
Good catch @danielgrant! I resolved my issue but actually not going the existing VPC route and instead peering the newly-created VPC with my existing VPC, which actually feels like a cleaner solution IMO. But this is still invaluable info. 👏

**Comment by vellichor** [comment]
We faced a very similar issue when deploying into an existing VPC with two route tables. In our case, the route tables were already existing when we created the cluster with kops, and kops did not apply the tag to either of them. In this case our error looked like

`E0311 14:22:50.957633       1 route_controller.go:120] Couldn't reconcile node routes: error listing routes: unable to find route table for AWS cluster: foo.k8s.local`

Manually applying the tag to the right routing table caused the cluster to recover with no more intervention.
